mitmproxy/proxy/layers/modes.py

```diff
@@ -128,17 +141,47 @@ def _handle_event(self, event: events.Event) -> layer.CommandGenerator[None]:
                 n_methods = self.buf[1]
                 if len(self.buf) < 2 + n_methods:
                     return
-                if SOCKS5_METHOD_NO_AUTHENTICATION_REQUIRED not in self.buf[2:2 + n_methods]:
-                    yield from self.socks_err("mitmproxy only supports SOCKS without authentication",
-                                              SOCKS5_METHOD_NO_ACCEPTABLE_METHODS)
-                    return
+
+                method = SOCKS5_METHOD_NO_AUTHENTICATION_REQUIRED
+                if self.auth_enable():
+                    if SOCKS5_METHOD_USER_PASSWORD_AUTHENTICATION not in self.buf[2:2+n_methods]:
+                        yield from self.socks_err("mitmproxy only support user password authentication",
+                                                  SOCKS5_METHOD_NO_ACCEPTABLE_METHODS)
+                        return
+                    method = SOCKS5_METHOD_USER_PASSWORD_AUTHENTICATION
 
                 # Send Server Greeting
-                # Ver = SOCKS5, Auth = NO_AUTH
-                yield commands.SendData(self.context.client, b"\x05\x00")
+                if method == SOCKS5_METHOD_NO_AUTHENTICATION_REQUIRED:
+                    # Ver = SOCKS5, Auth = NO_AUTH
+                    yield commands.SendData(self.context.client, b"\x05\x00")
+                    self.need_auth = False
+                else:
+                    # Ver = SOCKS5, Auth = UserPassword
+                    yield commands.SendData(self.context.client, b"\x05\x02")
+                    self.need_auth = True
                 self.buf = self.buf[2 + n_methods:]
                 self.greeted = True
 
+            if self.need_auth:
+                # Parse client authentication request
+                if len(self.buf) < 2:
+                    return
+
+                id_len = self.buf[1]
+                if len(self.buf) < 3 + id_len:
+                    return
+                pw_len = self.buf[2 + id_len]
+                if len(self.buf) < 3 + id_len + pw_len:
+                    return
+                test_id, test_pw = self.buf[2:(2 + id_len)], self.buf[(3 + id_len):(3 + id_len + pw_len)]
+                if self.check_auth(test_id, test_pw):
+                    yield commands.SendData(self.context.client, b"\x01\x00")
+                else:
+                    yield from self.socks_err("authentication failed", 0x01, ver_code=0x01)
```

Comment:
Why do you submit a custom `ver_code` here?
[RFC 1928](https://www.rfc-editor.org/rfc/rfc1928.html#section-6) always has this hardcoded to 0x05.

Reply:
I think this is a misunderstanding of the RFCs:
```
                         +----+--------+
                         |VER | METHOD |
                         +----+--------+
                         | 1  |   1    |
                         +----+--------+
```

The 1 in this table only indicates that the version field is one byte long. If my hypothesis is correct please confirm. No need to fix it, I'm working on the code now anyways :)